cpp/tests/stream_compaction/distinct_tests.cpp

```diff
@@ -138,3 +138,185 @@ TEST_F(Distinct, WithNull)
 
   CUDF_TEST_EXPECT_TABLES_EQUAL(expected_unequal, sorted_unequal->view());
 }
+
+TEST_F(Distinct, BasicList)
+{
+  using lcw = cudf::test::lists_column_wrapper<uint64_t>;
+  using icw = cudf::test::fixed_width_column_wrapper<cudf::size_type>;
+
+  // clang-format off
+  auto const idx = icw{ 0,  0,   1,      2,   1,      3,      4,   5,   5,      6,      4,     4 };
+  auto const col = lcw{{}, {}, {1}, {1, 1}, {1}, {1, 2}, {2, 2}, {2}, {2}, {2, 1}, {2, 2}, {2, 2}};
+  // clang-format on
+  auto const input = cudf::table_view({idx, col});
+
+  auto const exp_idx = icw{0, 1, 2, 3, 4, 5, 6};
+  auto const exp_val = lcw{{}, {1}, {1, 1}, {1, 2}, {2, 2}, {2}, {2, 1}};
+  auto const expect  = cudf::table_view({exp_idx, exp_val});
+
+  auto result        = cudf::distinct(input, {1});
+  auto sorted_result = cudf::sort_by_key(*result, result->select({0}));
+
+  CUDF_TEST_EXPECT_TABLES_EQUAL(expect, *sorted_result);
+}
+
+TEST_F(Distinct, NullableList)
+{
+  using lcw  = cudf::test::lists_column_wrapper<uint64_t>;
+  using icw  = cudf::test::fixed_width_column_wrapper<cudf::size_type>;
+  using mask = std::vector<bool>;
+
+  // clang-format off
+  auto const idx    = icw {  0,  0,   1,   1,      4,   5,   5,  6,       4,     4,  6};
+  auto const valids = mask{  1,  1,   1,   1,      1,   1,   1,  0,       1,     1,  0};
+  auto const col    = lcw {{{}, {}, {1}, {1}, {2, 2}, {2}, {2}, {}, {2, 2}, {2, 2}, {}}, valids.begin()};
+
+  auto const exp_idx    = icw {  0,   1,      4,   5,  6};
+  auto const exp_valids = mask{  1,   1,      1,   1,  0};
+  auto const exp_val    = lcw {{{}, {1}, {2, 2}, {2}, {}}, exp_valids.begin()};
+
+  // clang-format on
+  auto const input  = cudf::table_view({idx, col});
+  auto const expect = cudf::table_view({exp_idx, exp_val});
+
+  auto result        = cudf::distinct(input, {1});
+  auto sorted_result = cudf::sort_by_key(*result, result->select({0}));
+
+  CUDF_TEST_EXPECT_TABLES_EQUAL(expect, *sorted_result);
+}
+
+TEST_F(Distinct, ListOfStruct)
+{
+  // Constructing a list of struct of two elements
+  // 0.   []                  ==
```

Comment:
A few related questions about hashing behavior. Some of these may be answerable by reading the tests or implementation, but I want to ensure we're in sync about the intended behavior. Some of these may require additional test coverage to enforce equality / inequality.

1. Do empty lists hash to the same value for all child dtypes? e.g. `list<int> []` and `list<float> []`
2. Do lists of equivalent numbers of null values but different types hash to the same value? e.g. `list<int> [null]` and `list<float> [null]`
3. Do `struct<int> {null}` and `struct<float> {null}` hash to the same value? (non-null parent struct containing null child)

Reply:
Yes to all.
Since hash is always on a single table, the difference will come in the order of hashing of these columns. If we have a table that contains any of these pairs, a row containing `list<int> []`, `list<float> [1.0, ...]` will differ from another containing `list<int> [1, ...]`, `list<float> []` because in both cases, the `list<int>` is hashed first and then the subsequent is combined into it.

This behaviour is similar to the existing element_hasher where nulls hash to the same value regardless of type https://github.com/rapidsai/cudf/blob/ca952f8355810c46bb132f0b50a67e4109108ccb/cpp/include/cudf/table/row_operators.cuh#L435-L440

Reply:
I'm glad we have a precedent for this behavior to compare against, but it still seems odd to me. We usually rely on hashing as a cheap way to assess probable equality (equal hashes are necessary but not sufficient for equality). However, this behavior means there is a strong possibility of getting the same hash with table rows with different data types if they only contain nulls. Are there any algorithms that could compare hashes without first validating that the rows contain the same data types in the same order? If so, we'd want to break from that precedent of letting `table_row{float, int} = {null, null}` and `table_row{int, float} = {null, null}` have the same hash.

Reply:
This logic is very old. At the time, all I needed was that equivalent rows like `{1, "a", NULL}` and `{1, "a", NULL}` would generate the same hash. 

I suppose using a different `null_hash` per type makes sense, but was more effort than needed because there was never a case where I was working on tables with possibly different column types. Even if I had been, I doubt I would have cared because hashes are imperfect anyways. 

> Are there any algorithms that could compare hashes without first validating that the rows contain the same data types in the same order?

As a rule, I can all but guarantee that this will never happen, at least not directly in the implementation of any libcudf algorithm because anything working on two tables will always required the same types to avoid double type dispatching. And even if it did, I'm not all that concerned about the potential for more collisions. Furthermore, it raises philosophical questions like "Are two nulls of different types unequal?" that I'd rather leave buried.

Reply:
I'm perfectly satisfied with this answer! Thank you for the background and context -- it helps me as a reviewer to have a feel for what parts of the design are sharper/smoother (and whether they came to be that way by design or by historical accident). I can resolve this conversation.